determined-ai/determined
# feat: add gauge taiko e2e-tests

## Description

Replace [Cypress](https://www.cypress.io/) with [Gauge](https://gauge.org/) and [Taiko](https://taiko.dev/).

Motivations for this PR:
- cypress's wait element policy is only ok, resulting in a lot of timing flakes.
- cypress's wait policy for network calls is poor.
- cypress is unable to test beyond the initial browser tab.
- taiko's more reliable wait policy enables better performing and more reliable tests.
- taiko's approach to integration testing is more inline with a user's POV instead of approaching with a developer's mentality. Testing should satisfy user stories and not be framed and limited by development architecture decisions.

Core work supported in this PR:
- Add gauge and taiko
- Enable local testing with chromium (both with chrome and headless)
- Update determined cluster setup script to support gauge/taiko
- Enable lint check with gauge/taiko code
- Migrate cypress tests over to gauge/taiko
- Update CircleCI to test with gauge/taiko
  - Save test results
  - Save test artifacts such as reports and screenshots
  - Enable parallel spec testing to improve performance
- Remove cypress
- Update WebUI e2e testing documentation

Sample generated report accessible in Circle CI artifacts:
<img width="1419" alt="Screen Shot 2020-11-18 at 5 08 13 PM" src="https://user-images.githubusercontent.com/220971/99603530-bb6eea00-29c0-11eb-9dca-ee9fde901178.png">




## Test Plan

- ensure `test-e2e-webui` passes.
  - check `test-e2e-webui` job and check that `Tests` and `Artifacts` tab are populated with test results and test reports.
- check that `make dev-tests` and/or `make test` (headless mode) passes locally.




## Commentary (optional)




## Checklist

See [Release Note](../docs/release-notes/README.md) for details.





[DET-1234]: https://determinedai.atlassian.net/browse/DET-1234
